### PR TITLE
Fix #2634: Fix talkback bug on activity transition to topic screen

### LINKED ISSUE #2634: [A11y] Topic Info Tab
## Current Output

https://user-images.githubusercontent.com/9396084/107034739-24564900-67dd-11eb-9d87-b671a28ed6f5.mp4

## Issues Identified
The talkback reads this screen as **Oppia** at start, instead it should be **Topic: Fractions** (Assigned to @Sparsh1212)

Note: To understand the above issue(s) completely, it is recommended that you setup Talkback and play with the app keeping it on and that will give you better context.

**Accessibility Guide**: https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide

### PR DESCRIPTION
**Pull request by Sparsh1212**
## Explanation

Fixes #2634: The talkback previously used to read "Oppia" on transition to the topic info screen.  In this PR, this bug is fixed and it now reads "Topic: `topic_name`" instead of "Oppia".

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by Sparsh1212** [comment]
@rt4914 Please review.

File: app/src/main/AndroidManifest.xml
Comment (BenHenning) [inline app/src/main/AndroidManifest.xml]: @Sparsh1212 we should be setting an actual label here per #2634 (e.g. something that includes the activity title & the current topic open). That will require introducing a new string that supports a parameter (similar to https://github.com/oppia/oppia-android/blob/develop/app/src/main/res/values/strings.xml#L47).

Also, please make sure that you add a new test for this label to make sure it's being properly defined. This may require introducing a new ``TopicActivityTest`` in [this package](https://github.com/oppia/oppia-android/tree/develop/app/src/sharedTest/java/org/oppia/android/app/topic). See [TopicFragmentTest](https://github.com/oppia/oppia-android/blob/develop/app/src/sharedTest/java/org/oppia/android/app/topic/TopicFragmentTest.kt#L91) for an idea on how to properly set up ``TopicActivity`` for testing.

**Review by BenHenning — COMMENTED** [review]
Thanks @Sparsh1212! Left some thoughts.

**Comment by Sparsh1212** [comment]
Actually, how talkback works while transistion to any new `Activity`is that: it first checks the `label` defined in the `activity` tag of the corresponding Activity in `AndroidManifest.xml`. If there is no label present then it uses the value of `android:label` which is "Oppia" in our case. So because of this, the talkback is reading "Oppia" everytime we make a transition to any new Activity. And then it reads the `app_title`. So, in order to fix this, we have to explicitly set the `label` of the Activity as an empty string, which I did in my current PR.

Now, if we set the actual dynamic label (of the Topic) in the `activity` tag, then the talkback is reading the "Topic: `topic_name`" two times. This is happening because, one time it is reading due to the label of the `Activity` and the second time due to the AppBar `title`. Hence, we have to set the label as an empty string, in order to avoid this double reading.

Ref: [Link](https://stackoverflow.com/questions/20792904/how-do-you-get-a-newly-started-activity-to-announce-itself-in-android-talkback)

@rt4914 @BenHenning Please let me know your thoughts on this. Also, if we decide to set the label as an empty string then do I also have to add a new test for it ?

**Comment by rt4914** [comment]
This seems to be a nice issue and I am not able to find any straightforward solution for this. 
I have tried following things and the results were as follows:
1. Using https://stackoverflow.com/a/10138095 as reference I set the title before the activity is laid completely. This doe solve our issue but raises another issue. After this solution talkback reads, "Topic: Fractions", "Topic: Fractions". The first string is because of `label/title` for activity and second is because of `TextView` which is part of `Toolbar` in `TopicFragment`.
2. Following up on above solution, I made `TopicFragment` not important for accessibility and that reduced the first time output to only one string but later on the learner can never reach `Topic: Fractions` again.

@BenHenning Any thoughts?

**Comment by BenHenning** [comment]
Could we set the label of the activity to a static string like "topic page"? Then TalkBack would read "Topic page" - "Topic: Fractions" which seems more organizationally correct. It also better matches what sighted users see (which is generally how we should frame screenreader output so that the two experiences are closely aligned & special casing isn't needed as often).

@rt4914 @Sparsh1212

**Comment by Sparsh1212** [comment]
> Could we set the label of the activity to a static string like "topic page"? Then TalkBack would read "Topic page" - "Topic: Fractions" which seems more organizationally correct. It also better matches what sighted users see (which is generally how we should frame screenreader output so that the two experiences are closely aligned & special casing isn't needed as often).
> 
> @rt4914 @Sparsh1212

Yes, that will be more correct.

**Comment by rt4914** [comment]
@Sparsh1212 Can you make the changes as suggested by Ben. Use label as `Topic page`

**Comment by Sparsh1212** [comment]
> @Sparsh1212 Can you make the changes as suggested by Ben. Use label as `Topic page`

Sure, I will do the changes as suggested by Ben. Also, please tell me whether I have to write the test for it or not.

**Comment by rt4914** [comment]
> > @Sparsh1212 Can you make the changes as suggested by Ben. Use label as `Topic page`
> 
> Sure, I will do the changes as suggested by Ben. Also, please tell me whether I have to write the test for it or not.

@Sparsh1212 Yes.

**Comment by Sparsh1212** [comment]
@BenHenning Can you please be a little bit more descriptive about writing the test for the label? I am actually not able to understand what should be the exact purpose of this test. Like what all things should It test for in the label/title of the Activity?
Any clarification in this regard will be of great help to me.
Thanks.

**Comment by Sparsh1212** [comment]
@BenHenning Since, now I have to set a static string in the label, so do I need to write a test in this case or not? If yes, then please guide me on this.

**Comment by rt4914** [comment]
@Sparsh1212 Following things might help you.
One check this test: https://github.com/oppia/oppia-android/blob/8a893a295ee791135484f38dbb8925730c724994/app/src/sharedTest/java/org/oppia/android/app/topic/TopicFragmentTest.kt#L132

Similar to this test case you have to check for label/title.
You can get the title inside the test case using following code:
`val title  = activityTestRule.activity.title`
Once you get this title check this with string "Topic page" using `assertThat`.

**Comment by Sparsh1212** [comment]
@rt4914 Thankyou very much for guiding me.
I have made the changes as suggested by Ben and have also added a test for the label.
@BenHenning PTAL.

File: app/src/main/AndroidManifest.xml
Comment (rt4914) [inline app/src/main/AndroidManifest.xml]: Shift this to above line.
```
android:label="@string/topic_page" />
```

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Use string from `strings.xml` file using `R.string.topic_page`

**Review by rt4914 — CHANGES_REQUESTED** [review]
@Sparsh1212 Suggested changes.

**Comment by Sparsh1212** [comment]
> @Sparsh1212 Suggested changes.

Done. PTAL, thanks.

**Comment by rt4914** [comment]
@Sparsh1212 Please reply to the comments otherwise its assumed that you have not solved yet.
Also update your branch with latest develop so that all checks can pass.

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Check `AdminstratorControlsActivityTest` to see how to use `context` easily.

**Review by rt4914 — CHANGES_REQUESTED** [review]
PTAL

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (Sparsh1212) [reply app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Done.

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (Sparsh1212) [reply app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Done.

File: app/src/main/AndroidManifest.xml
Comment (Sparsh1212) [reply app/src/main/AndroidManifest.xml]: Done. PTAL

**Comment by Sparsh1212** [comment]
> @Sparsh1212 Please reply to the comments otherwise its assumed that you have not solved yet.
> Also update your branch with latest develop so that all checks can pass.

Sure, I will remember this from the next time. I have done all the changes. PTAL

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Remove this blank line.

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt:70]: Add blank line above this

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Remove this. We don't need this in this test.

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Remove this.

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Remove.

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Remove this entire function

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: While merging with develop you have missed `FakeOppiaClockModule` in these modules and because of which tests are failing.

Update this line to:
`FirebaseLogUploaderModule::class, FakeOppiaClockModule::class`

**Review by rt4914 — CHANGES_REQUESTED** [review]
@Sparsh1212 Once the suggested changes are done, your tests should pass.

**Comment by Sparsh1212** [comment]
> @Sparsh1212 Once the suggested changes are done, your tests should pass.

Okay, thank you very much. I'll do the suggested changes ASAP.

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (Sparsh1212) [reply app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Updated

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (Sparsh1212) [reply app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Removed

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (Sparsh1212) [reply app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Removed

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (Sparsh1212) [reply app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Removed

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (Sparsh1212) [reply app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Removed

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (Sparsh1212) [reply app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt:70]: Added

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (Sparsh1212) [reply app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Removed

**Comment by Sparsh1212** [comment]
@rt4914 PTAL, I have made all the suggested changes and now the build tests have also passed. Thankyou very much for guiding me.

**Review by rt4914 — APPROVED** [review]
LGTM. You will need to again merge with develop but lets wait for @BenHenning review and once you get approval from him too then update your branch so that we can merge.

Thanks.

**Comment by Sparsh1212** [comment]
@BenHenning PTAL thanks.

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: ``testTopicActivity_hasCorrectActivityLabel`` might be a bit more succinct.

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Prefer fitting this on one line if it fits within the column limit.

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: This is essentially a change detector test which I think is fine in this case since it's a good proxy for validating the label TalkBack will read in different circumstances. I suggest adding a comment explaining the purpose of this test somewhere, e.g. before the assertion line you could add:

```kotlin
// Verify that the activity label is correct as a proxy to verify TalkBack will announce the correct string when it's read out.
assertThat(...)
```

(Note that the comment might need to be line-wrapped to fit within the 100 column limit).

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @Sparsh1212! Just had 3 comments, otherwise I think this looks good.

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (Sparsh1212) [reply app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Done

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (Sparsh1212) [reply app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Done

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (Sparsh1212) [reply app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Done

**Comment by Sparsh1212** [comment]
@rt4914 @BenHenning I have made the suggested changes and have also updated with the latest develop branch. PTAL

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]:
```suggestion
    // Verify that the activity label is correct as a proxy to verify TalkBack will announce the
    // correct string when it's read out.
```

While a block comment is perfectly fine, I suggest using single-line comments for broader consistency. In general we only use block comments for explaining large paragraphs of text. Note also how the comment is written to fill up as much of the line as possible--prefer this versus breaking up multiple lines much earlier than the column limit.

**Review by BenHenning — COMMENTED** [review]
Thanks @Sparsh1212! Just one comment. PTAL.

File: app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt
Comment (Sparsh1212) [reply app/src/sharedTest/java/org/oppia/android/app/topic/TopicActivityTest.kt]: Okay, I will keep this in mind the next time. Now I have updated the comment will single-line comments. PTAL

**Comment by Sparsh1212** [comment]
@BenHenning Can you please take a look again. Thanks.

**Review by BenHenning — APPROVED** [review]
Thanks @Sparsh1212! LGTM. Apologies for the delay--I often skip code reviews on weekends. I probably shouldn't have marked this as 'requested changes' in that case--will try to be more mindful about this in the future.

**Pull request merged**
